File: 142_Linked_List_Cycle_II/142_2.py
class Solution:
    def detectCycle(self, head):
        if not head or not head.next:
            return None
        if head == head.next:
            return head
        if head == head.next.next:
            return head
        fast, slow = head, head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                break
        if not fast or not fast.next:
            return None
        slow = head
        while slow != fast:
            fast = fast.next
            slow = slow.next
        return slow

File: 142_Linked_List_Cycle_II/142_2.java
public class Solution {
    public ListNode detectCycle(ListNode head) 
    {
        if (head == null || head.next == null) return null;
        if (head.next == head) return head;
        if (head.next.next == head) return head;
        ListNode fast = head;
        ListNode slow = head;
        while (fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) break;
        }
        if (fast == null || fast.next == null) return null;
        slow = head;
        while (slow != fast)
        {
            fast = fast.next;
            slow = slow.next;
        }
        return slow;
    }
}